docs/design/guideline_recategorization.md

```diff
@@ -0,0 +1,178 @@
+                    }
+                }
+            }
+        }
+    },
+    "required": [],
+    "type": "object"
+}
+```
+
+The second validation mechanism is the generation of a `guideline-recategorization-plan-report.md` containing alerts on semantically incorrect recategorizations.
+That is, possible recategorizations that are not described as valid in the introduction.
+Semantically invalid recategorizations will be detected by looking at a query’s categorization and its effective categorization (i.e., its applied recategorization).
+
+In addition, an update to the `deviations_report.md` report’s invalidate deviations table will provide feedback to users that apply deviations to guidelines with an effective category equal to `mandatory` which cannot be deviated from.
+The changes to generate the new report and update the existing report will be made in the report generation script `scripts/reports/analysis_report.py`.
+
+### SARIF rewriting
+
+The *effective category* of a guideline is a runtime property that is not reflected in the SARIF result file and therefore is not visible in any viewer used to view the results (e.g., [Code Scanning](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/about-code-scanning)).
+To ensure that users can view the *effective category* we need to rewrite the `@tags` query metadata property in the SARIF result file.
+The `@tags` value is a JSON array located at the [JSON path](https://datatracker.ietf.org/wg/jsonpath/about/):
+
+`$.runs[?(@.tool.driver.name="CodeQL")].tool.driver.rules[*].properties.tags`
+
+The category tag has the form `external/<standard>/obligation/<category>`
+Each guideline has an `external/<standard>/id/<rule-id>` tag that can be used to determine if a recategorization is applicable by performing a case insensitive compare on the `<rule-id>` extracted from the query’s tags array and the value of the rule-id key in a `guideline-recategorizations` section.
+The rewriting replaces the `<category>` part in `external/<standard>/obligation/<category>` with the newly specified category and adds a new tag `external/<standard>/original-obligation/<category>` with the rule’s original category.
+
+The rewrite process translates each entry in the guideline recategorization specification into a [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902) specific to the processed SARIF file. The JSON Patch is SARIF file specific due to its reliance on [JSON Pointer](https://www.rfc-editor.org/rfc/rfc6901) to locate the obligation tags.
+
+A new SARIF file is created by applying the JSON Patch to the processed SARIF file.
+
+## Non-MISRA standards
+
+Guideline recategorization applies to rules adhering to the MISRA categorizations.
+For standards that deviate from these conventions we assume that the rules have an *effective category* equivalent to MISRA’s *required* category.
```

Comment:
```suggestion
For standards that deviate from these conventions we assume that the rules have an *effective category* equivalent to MISRA’s *required* category.
CERT rules, for example, are handled in the same way as MISRA's required rules.
```

Reply:
When reading this, do you expect to be able to recategorize CERT rules to _mandatory_?

Reply:
I think mandatory is a pretty dumb category of rules, so I would say no.  I don't like saying that absolutely under every condition we are right and there is absolutely no use case we haven't considered.